multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/RestartAppStep.java

```diff
@@ -95,7 +97,22 @@ protected List<AsyncExecution> getAsyncStepExecutions(ProcessContext context) {
 
     @Override
     public Duration getTimeout(ProcessContext context) {
-        return context.getVariable(Variables.START_TIMEOUT);
-    }
+        CloudApplicationExtended app = context.getVariable(Variables.APP_TO_PROCESS);
+        Integer startTimeout = extractUploadTimeoutFromAppAttributes(app, SupportedParameters.START_TIMEOUT);
+        Duration startTimeoutOperational = context.getVariable(Variables.START_APP_TIMEOUT);
+
+        Duration resultTimeout;
+        if (startTimeoutOperational != null) {
+            resultTimeout = startTimeoutOperational;
+        } else if (startTimeout != null) {
+            resultTimeout = Duration.ofSeconds(startTimeout);
+        } else {
+            int startTimeoutGlobal = (int) context.getVariable(Variables.START_APP_TIMEOUT_GLOBAL)
+                                                  .toSeconds();
+            resultTimeout = Duration.ofSeconds(startTimeoutGlobal);
+        }
 
+        logTimeout(Messages.START_APP_TIMEOUT, resultTimeout.toSeconds());
```

Comment:
Can you extract this logic in a method that can be reused for the other steps. maybe you can place it in TimeoutAsyncFlowableStep